src/frontend/src/handler/create_table.rs

```diff
@@ -258,10 +258,16 @@ pub fn bind_sql_column_constraints(
                 ColumnOption::DefaultColumns(expr) => {
                     let idx = binder
                         .get_column_binding_index(table_name.clone(), &column.name.real_value())?;
-                    let expr_impl = binder.bind_expr(expr)?;
+                    let mut expr_impl = binder.bind_expr(expr)?;
 
                     check_default_column_constraints(&expr_impl, column_catalogs)?;
 
+                    let target = column_catalogs[idx].data_type().clone();
+
+                    if expr_impl.return_type() != target {
+                        expr_impl = expr_impl.cast_assign(target)?;
+                    }
```

Comment:
This is checked in `new_cast`. We can always call `cast_assign`.

https://github.com/risingwavelabs/risingwave/blob/ca18ca82fca480202c4d5dfed2af529d1efcf5ca/src/frontend/src/expr/function_call.rs#L145-L147